[docs] Edits on new Styles docs (#503)
Polymer/lit-element: docs/_guide/styles.md

LitElement lets you define static styles that apply to all instances of a component.

We recommend using static styles for optimal performance. LitElement uses [Constructable Stylesheets](https://wicg.github.io/construct-stylesheets/) in browsers that support it, with a fallback for browsers that don't. Constructable Stylesheets allow the browser to parse styles exactly once and reuse the resulting Stylesheet object for maximum efficiency.

The styles in the static `styles` property are evaluated once only and applied to all instances of the element. You can modify styles per element instance by using [CSS custom properties](#cssprops):

*   The element **developer** defines a style rule inside shadow DOM that uses CSS variables to apply a custom CSS property to a particular style property.
*   The element **consumer** defines the value of the custom CSS property outside shadow DOM in a CSS rule that is inherited by the host element.

If you don't want to use custom properties, you can define per-instance styles in a `<style>` element inside shadow DOM. See the section on [Defining your styles in a style element](#styleelement) for more information.

> </div>

Reviewer: <div> already closed in line 15 :)

  @property() mainColor = 'blue';
  render() {
    return html`
      <style>
        :host {
          color: ${this.mainColor};
        }
      </style>
    `;
  }
}
```

### Define styles in an external stylesheet {#external-stylesheet}

We strongly recommend static styles, CSS custom properties, or [lit-html's `classMap` or `styleMap` directives](TODO) if you're stying non-host shadow root contents.

However, you can load an external stylesheet into a shadow root with a `<link>` element:

```js
import {LitElement} from 'lit-element';

class MyElement extends LitElement {
  render() {
    return html`
      <link rel="stylesheet" href="./styles.css">
    `;
  }
}
```

This can be a good way to load CSS generated from tools like SASS/LESS.

There are some important caveats though:

* External styles can cause a flash-of-unstyled-content (FOUC) while they load.
* The URL in the `href` attribute is relative to the _main document_, making this technique mostly useful for application elements where asset URLs are well known, and not for reusable elements published publicly.

### Write CSS styles for a host element and its shadow DOM {#shadow-dom-styles}

In this section: 

> * [Write CSS styles for a host element](#host-styles). The host element is your component. It is called the "host" because it hosts shadow DOM. You can style the element itself with the special CSS selectors `:host` and `:host()`.

Reviewer 2: description duplicates the into to this page

* [Write CSS styles for elements in shadow DOM](#shadow-dom-styles).

> * [Write CSS styles for slotted children](#slotted-styles). Light DOM children of your host element can be rendered in shadow DOM via the `slot` element. You can style slotted content with the `::slotted()` CSS selector.

Reviewer 2: Maybe leave off the second sentence, and change the first to just define "slotted children", like:

> Slotted elements are light DOM children of the host that are rendered into a shadow root with the `<slot>` element.

slug: styles
---

{::options toc_levels="1..3" /}
* ToC
{:toc}

## Overview

This page describes how to style a web component built with LitElement and using shadow DOM.

By default, LitElement creates a shadow root and renders into [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM).

Shadow DOM scopes CSS so that styles defined in a shadow root only apply to DOM inside the shadow root, and do not "leak" to outside DOM. Shadow roots are also isolated from styles defined outside the shadow root, whether in the main page or an outer shadow root.

**This guide applies only if you use the default (shadow DOM) render root.** If you [modify your element's render root](templates#renderroot) to render into the main DOM tree instead of a shadow root, these instructions won't apply.

> {.alert-info}

Reviewer 2: Staging site shows this isn't correct. oops

> {.alert-info}

Author: nah that doesn't work, i've been doing `<div class="alert alert-whatever">`

**This guide applies only if you use the default (shadow DOM) render root.** If you [modify your element's render root](templates#renderroot) to render into the main DOM tree instead of a shadow root, these instructions won't apply.
{.alert-info}

**If you're using the Shady CSS polyfill, be aware that it has some limitations.** See the [Shady CSS README](https://github.com/webcomponents/shadycss/blob/master/README.md#limitations) for more information.
{.alert-info}

## Developing a LitElement component

### Where to put your styles

When developing a component there are 3 main places you can define its styles:

* In the static `styles` property of a LitElement class (recomended).
* Inline, inside a `<style>` element within the HTML template defined in your `render` function.
* In an external stylesheet, linked from your LitElement template with `<link rel="stylesheet" href="...">`.

In addition, some CSS properties are inherited. Inheritance does flow down the shadow DOM tree.

#### Define styles in a static styles property

LitElement lets you define static styles that apply to all instances of a component.

<div class="alert alert-info">

> **Use static styles for optimal performance.** LitElement uses [Constructable Stylesheets](https://wicg.github.io/construct-stylesheets/) in browsers that support this new standard, with a fallback for browsers that don't. Constructable Stylesheets allow LitElement to parse styles exactly once and reuse the resulting Stylesheet object for maximum efficiency.

Reviewer 2: ```suggestion
**Use static styles for optimal performance.** LitElement uses [Constructable Stylesheets](https://wicg.github.io/construct-stylesheets/) in browsers that support this new standard, with a fallback for browsers that don't. Constructable Stylesheets allow the browser to parse styles exactly once and reuse the resulting Stylesheet object for maximum efficiency.
```

</div>

> SStatic styles are evaluated once only, they apply to all instances & never re-evaluated. To define styles per instance, we recommend CSS custom properties. See [theming](#theming). Or you cam use inline styles. See [inline styles](#inline).

Reviewer 2: ```suggestion
Static styles are evaluated once only, they apply to all instances and never re-evaluated. To define styles that change per instance, we recommend CSS custom properties. See [theming](#theming). Or you can use inline styles. See [inline styles](#inline).
```

**If you're using the Shady CSS polyfill, be aware that it has some limitations.** See the [Shady CSS README](https://github.com/webcomponents/shadycss/blob/master/README.md#limitations) for more information.
{.alert-info}

> ## Developing a LitElement component

Reviewer 2: "Developing" -> "Styling"?

Author: i want to specifically call out "developing" vs "consuming" - when you're developing a component you want to know how to mess around with shadow DOM styles etc, but when you're consuming, this should all be  a black box and you should just be styling my-element {...} from wherever you import it

Author: Yep I see that. Will think about it & update accordingly - thanks both!

}
```

> The `unsafeCSS` tag is for...?

Reviewer 2: We can leave it out for now and add it later.

> #### Define styles inline in a style block

Reviewer 2: "style block" isn't a well-defined term, IMO. "style element"?

> You can also style a shadow root by including inline styles in your element template. We still recommend static styles, but in some cases you may want to vary the CSS per-element. One way to do this is with bindings in `<style>` elements. 

Reviewer 2: "element's template"?

}
```

TODO: The `unsafeCSS` tag is for...?

sorvell: The intention of unsafeCss is to allow you to use any value. To be clear, the function is named this way so that it's clear that you may be incurring some risk.

See also https://github.com/Polymer/lit-element/issues/488 

> #### Define styles inline in a style block {#inline}

Reviewer 2: Do we have a way to refer to `<style>` tags in the template, whether or not they have bindings? "External" refers to `<link>`, so... "internal"?

}
```

> TODO: Does changing `this.mainColro` reevaluate and restyle dynamically whenever `render()` is called? (my tests suggest no)

Reviewer 2: It absolutely should.

> #### Define styles in an external stylesheet

Reviewer 2: ```suggestion
#### Load styles in an external stylesheet
```

"load" because the dev might be using someone else's styles. The critical bit is how they're loaded.

There are some important caveats though:

* External styles can cause a flash-of-unstyled-content (FOUC) while they load.
* The URL in the `href` attribute is relative to the _main document_, making this technique mostly useful for application elements where asset URLs are well known, and not for reusable elements published publicly.

### What you can style

You can style:

> * [The host element](#host). The host element is your component. It is called the "host" because it hosts shadow DOM. You can style the element itself with special CSS selectors `:host` and `:host()`.

Reviewer 2: "component": How consistently are we referring to "element" vs "component"? I've been trying to use "element" since we're always talking about HTML elements.

* [Elements in the host's shadow root](#children).

> * [Slotted elements](#slot). Slotted elements are rendered in shadow DOM via the `slot` element - see MDN for more info. You can style slotted content with `::slotted()`.

Reviewer 2: ```suggestion
* [Slotted elements](#slot). Slotted elements are light DOM children rendered in shadow DOM via the `slot` element - see MDN for more info. You can style slotted content with `::slotted()`.
```

Added "light dom children" event though that's not completely accurate because of re-projection, because I think the current phrasing doesn't differentiate from normal shadow DOM content. As far as most users are concerned (I think), the important part is that this is how you style children.

{% include project.html folder="docs/style/hostselector" openFile="my-element.js" %}

COMPLETE NONSENSE, CORRECTED IN SEPARATE PR 

------ start nonsense -----

> 

Reviewer 2: I think the specificity bit is sensical, and important :)

Author: it's wrong tho, think i got it backwards

See [Custom Element Best Practices](https://developers.google.com/web/fundamentals/web-components/best-practices) for more information.

> #### Styling elements in the host's shadow root {#children}

Reviewer 2: do we really need "the host's" here? what else would be in the shadow root?

> #### Styling elements in the host's shadow root {#children}

Reviewer 2: I think `#children` is the wrong fragment. Children should refer to light DOM children.

To style elements in a shadow root, simply use standard CSS selectors.

Since CSS selectors in a shadow root only apply to elements in the shadow root, you don't need to be defensive against accidentally styling other elements in the page. This means you can generally write much simpler selectors, that are easier to reason about, and faster, than without shadow DOM.

> Simple selectors, like `*`, tagname, id and class selectors, are fine in a shadow root because they don't match outside the root:

Reviewer 2: ```suggestion
Simple selectors, like `*`, tagname, id, and class selectors, are fine in a shadow root because they don't match outside the root:
```

}
```

#### Styling slotted elements {#slotted}

> Use the `::slotted()` CSS pseudo-element to select light DOM elements that have been included in shadow DOM via the `<slot>` element.

Reviewer 2: ```suggestion
Use the `::slotted()` CSS pseudo-element to select light DOM children that have been included in shadow DOM via the `<slot>` element.
```

"children" because I suspect it's a bit more inline with intuition here. If we can explain or link to projection somewhere (maybe in the "what you can style" part), that might help.
